### DISCUSSION TITLE
Succint way to say "either one of or both"

### DISCUSSION THREAD

**Discussion by mickh** [metosin/malli#880] [Q&A]
Context: I want to spec a contact map, which can contain either an email address, a phone number, or both. But containing neither is invalid.

I currently have:
```
[:or
 [:map
  [:email :email]]
 [:map
  [:phone :phone-number]]
 [:map
  [:email :email]
  [:phone :phone-number]]]
```
which is not very DRY. I'm guessing that I can do better. Any help appreciated.

I know I can use a fn predicate, as in:
```
[:and
 [:map
  [:email {:optional true} :email]
  [:phone {:optional true} :phone-number]]
 [:fn not-empty]]
```
but am wondering if I can do it without using a fn predicate.

thanks

**Comment by NoahTheDuke** [comment]
Could use a local registry. Then the non dry one is much shorter.

**Reply by mickh** [reply]
Not sure I follow. How would a local registry help?

**Reply by NoahTheDuke** [reply]
I may have misunderstood your problem. I was thinking that `:email` and `:phone-number` are standing in for much longer schemas and the issue is the repetition of the types. But that's not your actual issue, so never mind.

**Comment by ikitommi** [comment]
Maybe just mark the keys as optional?

**Reply by ikitommi** [reply]
responded to the original message from email notification, yes, the `:fn` + `:and` is a ok way to check no non-emptiness.

**Reply by mickh** [reply]
Thanks. I was looking to see if there's a way to do it declaratively, but maybe there isn't. Too edge-casey?

**Reply by ikitommi** [reply]
not easy to do simpler than what is available already I think.

**Comment by ikitommi** [comment]
there is an existing discussion in this issue: https://github.com/metosin/malli/issues/474

**Reply by mickh** [reply]
Thanks for the link.

**Comment by kinthaiofficial** [comment]
"Either one of or both" — this is inclusive-or / disjunction, which malli expresses as `[:or ...]` but with each branch capturing partial data differently. Here's how to handle the common cases:

**Case 1: Map with one or both optional keys** (but at least one must be present)
```clojure
[:and
 [:map
  [:a {:optional true} :string]
  [:b {:optional true} :string]]
 [:fn (fn [{:keys [a b]}] (or a b))]]
```
The `[:fn ...]` validator enforces that at least one is present.

**Case 2: Inline merging with malli's `:merge`**
```clojure
[:or
 [:map [:a :string]]
 [:map [:b :string]]
 [:map [:a :string] [:b :string]]]
```
Explicit but verbose — you list all valid combinations.

**Case 3: Using `[:maybe ...]` for each field with semantic constraint**
```clojure
(def at-least-one-schema
  [:and
   [:map [:a [:maybe :string]] [:b [:maybe :string]]]
   [:fn {:error/message "at least one of :a or :b is required"}
    (fn [{:keys [a b]}] (or a b))]])
```

**For generating test data** (when using malli with generators), the `:fn` validator makes generation harder. If you need generation, enumerate the cases explicitly with `[:or ...]`.

**What this typically represents**:
This pattern appears often in optional configuration (either endpoint A or endpoint B or both), contact info (phone or email or both), etc.

Are you looking for validation only, or also generation/transformation with the schema?
